### ISSUE TITLE
Improve UX around read/write/forms panel management

### ISSUE THREAD

**Issue by marbemac** [stoplightio/studio#161]
**User story.**

As a user, I can quickly understand the ui panel layout, and it remains consistent so that I be more efficient.

**Is your feature request related to a problem?**

In general, there is a feeling of “jumpiness”. If you have Left: Read, and Right: Write, clicking on Write on the Left just swaps the two forms. It is a decision on our part to not show the same panel twice, but it is confusing to have the right panel change when the left panel does.

This also leads to another bit of confusion - when i click on a model, i don't know which view it will load, or on which side. 

**Describe the solution you'd like**

Follow standard editor conventions a bit more. When you open a file, you get ONE panel. There is always one panel unless you chose a second panel. The left is always about “Doing” (Write, or Forms if the active thing supports forms editing) and the right is the preview if the active node (file, operation, model, etc) supports it.

Note that this removes need for the fullscreen toggle altogether.

The primary compromise being made here: cannot display the code and forms panels side by side - must toggle between them. Can, however, still display the read/preview panel next to the editing (either code or forms) panel.

**Questions**

1. Location of the preview toggle? One option is bottom right, next to mocking and linting (basic mockup pictured below). Could then update mocking and linting tabs to use entire second panel area - compromise is cannot have read and linting open at same time, but not sure if that's a huge deal. Open to suggestions.
2. What to do on smaller screens, when it's not practical to have two panels side by side? Does preview panel cover the editing panel when user opens it?
3. If the active node does not support forms, do we not display the little mode toggle at all to keep things clean (the button toggle in top right of panels), or show it with code active and the forms option disabled? For example when looking at a markdown file, could keep things cleaner by not showing the toggle at all.
4. Potential language update:
  - Raw, Code, or Write?
  - Form or Forms?
  - Read or Preview?

<img width="898" alt="Screen Shot 2019-10-02 at 9 01 11 PM" src="https://user-images.githubusercontent.com/847542/66093943-c9ad3d00-e557-11e9-9d6a-b098eb6ec465.png">

---

**Final layout**:
- edit panel, two modes: Form or Code
- preview panel can be toggled on / off
- panel action bar above panels and below top bar
- for small screens, edit mode / preview toggle combined into a single three mode toggle
- full screen button at panel action bar - full screen covers top, side and meta bars - literally full screen
- switching to node without preview - disable preview button and use only edit panel. If preview was active before it stays active (in terms of app state) so if switched to node supporting preview it would be visible
- switching from form to node without forms - show write mode (if available). Layout settings remain so if switched back to node supporting form - it is still active
- switching to node supporting only preview (image) - activate preview mode
- if only write/form is available - don't show edit buttons

![Peek 2019-11-20 22-15](https://user-images.githubusercontent.com/14196079/69246382-972cd180-0be3-11ea-9438-cd95cba11122.gif)

**Comment by philsturgeon** [comment]
1. Personally I think bottom right is basically hidden. A few folks have already missed they did not see the mock server at all. Maybe more VS Code copying, where its one of a few options on the left? 

<img width="123" alt="image" src="https://user-images.githubusercontent.com/67381/66119239-4a932580-e5a6-11e9-98e0-31a6f4469e25.png">

2. Smaller screens could can just be cramped with two panels, and offer other ways to save space. When im "getting to work" on something I dont always need the navigator, so I hide that away. 

<img width="462" alt="image" src="https://user-images.githubusercontent.com/67381/66119675-1835f800-e5a7-11e9-9535-54db334bbe5b.png">

3. If the file open in general has various modes, they could be in the toolbar. When they are unavailable in a specific context (wrong line) they can be grey.

4. dunno, help!

**Comment by marbemac** [comment]
> Personally I think bottom right is basically hidden. A few folks have already missed they did not see the mock server at all. Maybe more VS Code copying, where its one of a few options on the left?

Wellll vscode is actually more hidden IMHO. It's the little icon with the magnifying glass on the right side of the screenshot below.

Introducing an entirely new base layout ui element (a left docked gutter) is a big decision that I think we should punt to later :). Ideally we can get this issue done within the constraints of our current layout.

Another option is a button toggle with forms/write and preview on right side of top navbar? Related to your comment for in item 3.

**Comment by mallachari** [comment]
Another option could be having a split screen button (like the one in vs code) next to form/write switch in toolbar but instead of doubling current panel it would open read view if it's available. For read panel there could be just close button in panel toolbar. 

For smaller screens there could be the same behavior as it is now which is switching between write/form/read - that shouldn't be confusing

**Comment by philsturgeon** [comment]
@marbemac What is a left docked gutter? I meant: the button that enables it is hard to spot, so put the button somewhere more obvious. If you are talking about where the loaded content goes and what it is, I think this is a bit small for larger APIs, and possibly a bit confusing for multiple API situations. 

<img width="592" alt="image" src="https://user-images.githubusercontent.com/67381/66202480-2454af00-e6a6-11e9-8292-ee1997954c84.png">

This could be its own page. So "Editor" is one thing, but "Mock" is another thing. Right now this would literally just mean moving that panel to be a full screen instead of having it smushed in the corner, but over time this would be rather helpful for when we add more functionality to Prism like tracking requests, etc. We could even have the same "read on the right" option, to see docs and Try it Now. 

None of this is new functionality, just moving what we already have around a bit. :)

**Comment by philsturgeon** [comment]
I was just getting that thought down but immediately after posting I figure: we should completely ignore all mocking comments and get that into another ticket. This can just be about always having Edit (Form, Text) on full screen or left, and always having "Preview" or "View" on the right.

**Comment by marbemac** [comment]
Here's a wire that puts the panel UX in a pretty prominent spot (top bar). Also highlights the fact that it's forms OR code, not both at same time - this is the big compromise to the approach being discussed in this issue - a little less flexible, but easier both technically and more intuitive UX wise.

<img width="1419" alt="Screen Shot 2019-10-10 at 10 38 09 PM" src="https://user-images.githubusercontent.com/847542/66622775-a1a08800-ebae-11e9-8941-b4e8342f71f6.png">

**Comment by marbemac** [comment]
- edit panel, two modes: Form or Code
- preview panel can be toggled on / off
- for small screens, edit mode / preview toggle combined into a single three mode toggle
- leave full screen toggle button on the panels themselves (fullscreen now goes over sidebar)
- best effort on the mode toggle? e.g. if form active and go to node that does not support form, we'll end up rendering code, but disable edit toggle with on hover message indicating current node does not support form

- TODO: marc to provide more info on top bar button organization (we have a lot right now)

**Comment by chris-miaskowski** [comment]
Hey "gang" ;) 

It's great to have all that conversation and requirements but... the story description, Marc's last comment, and the GIF in the Pull Request say three different things.

And the reality is actually different than any of the above :P 

I'm still slightly confused about the expected end result. Can you elaborate? Here is what's on staging:

![studio-panels-ui.gif](https://images.zenhubusercontent.com/5948db0651aa2c3057822fd0/f82608f9-f4bc-4af1-8b39-28d1e683770e)

I would very, very much appreciate if issues were updated with the final solution. It will save everyone an incredible amount of time in the long run. @mallachari can you please remember, as the person implementing it, to update the description?

**Comment by mallachari** [comment]
Yes, it does look ok. It looks different than the one from PR as it was done separately of this story. I mean Marc updated it and changed a bit after this story was done and I think it was just a single PR that wasn't referenced here.
So the one from PR is how I made it, and the current one is how it was updated later. 

I agree that the description here could be updated with current version to avoid confusion - I'll add it.
